api/server/router/volume/volume_routes.go

```diff
@@ -159,25 +159,35 @@ func (v *volumeRouter) deleteVolumes(ctx context.Context, w http.ResponseWriter,
 	}
 	force := httputils.BoolValue(r, "force")
 
-	version := httputils.VersionFromContext(ctx)
-
+	// First we try deleting local volume. The volume may not be found as a
+	// local volume, but could be a cluster volume, so we ignore "not found"
+	// errors at this stage. Note that no "not found" error is produced if
+	// "force" is enabled.
 	err := v.backend.Remove(ctx, vars["name"], opts.WithPurgeOnError(force))
-	// when a removal is forced, if the volume does not exist, no error will be
-	// returned. this means that to ensure forcing works on swarm volumes as
-	// well, we should always also force remove against the cluster.
-	if err != nil || force {
+	if err != nil && !errdefs.IsNotFound(err) {
+		return err
+	}
```

Comment:
The conditional flow here (and through all of these blocks) feels very clunky. Is there a better way to structure this that makes the flow more obvious, even if we have mode code movement?

Reply:
Yes, I was hoping to remove the "force" from the backend entirely, and just handle `notfound` errors (which are the ones we want to ignore) to the caller site; https://github.com/moby/moby/pull/45148#discussion_r1133268416

Unfortunately, there is code in the backend that still needs it (if the _plugin_ doesn't have the volume, then our local storage still has to remove it, and not fail)

_Perhaps_ still possible, if we would have the backend ignore the error, for what it needs to do, but make it still return it, but that still requires more changes (probably out of scope for this PR)